File: 백준/Gold/4256. 트리/트리.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

setrecursionlimit(10**9)
input = stdin.readline


def postorder(preorder, inorder,ans):
    if not inorder:
        return
    root = preorder.pop(0) #  전위 순회의 가장 앞은 항상 root 노드 
    idx = inorder.index(root) # 중위 순회는 root를 기준으로 왼,오가 나뉜다
    postorder(preorder, inorder[:idx],ans) # left subtree
    postorder(preorder, inorder[idx + 1:],ans) # right subtree
    ans.append(root)
    

    

tc = int(input())

for _ in range(tc):
    n = int(input())
    preorder = list(map(int,input().split()))
    inorder = list(map(int,input().split()))

    ans = []
    postorder(preorder,inorder,ans)
    print(*ans)
    
    
    


    




    

File: 백준/Gold/4256. 트리/트리.swift
import Foundation

let tc = Int(readLine()!)!

for _ in 0..<tc {
    
    let n = Int(readLine()!)!
    
    var preOrder = readLine()!.split{ $0 == " " }.compactMap{ Int(String($0))! }
    let inOrder = readLine()!.split{ $0 == " " }.compactMap{ Int(String($0))! }
    var result:[String] = []
    postOrder(preOrder: &preOrder, inOrder: inOrder)
    
    print(result.joined(separator: " "))
    
    func postOrder(preOrder: inout [Int], inOrder:[Int]){
        
        if inOrder.isEmpty || preOrder.isEmpty {
            return
        }

        let root = preOrder.first ?? 0
        preOrder.remove(at: 0)
       
        let index = inOrder.firstIndex(of: root) ?? 0
        postOrder(preOrder: &preOrder, inOrder: Array(inOrder[0..<index]))
        postOrder(preOrder: &preOrder, inOrder: Array(inOrder[index+1..<inOrder.count]))
        result += [String(root)]
    }
}






